File: source/hc_rust/core/source/kernel/lexer.rs
use super::lookup_table::CHAR_LU_TABLE;
use super::utf8::*;

/////////////////////////////////////////////
// LEXER
/////////////////////////////////////////////

#[derive(Debug, Copy, Clone)]
pub struct Lexer {
    pub byte_offset: u32,
    pub token_offset: u32,
    pub token_length: u32,
    pub byte_length: u32,
    pub prev_byte_offset: u32,
    pub prev_token_offset: u32,
    pub peek_byte_offset: u32,
    pub peek_token_offset: u32,
    pub token_type: i32,
    pub line: u32,
    pub input: &'static [u8],
}

impl Lexer {
    pub fn new(input_buffer: *const u8, input_len_in: usize) -> Lexer {
        use std::slice;

        let u8_slice: &[u8] = unsafe { slice::from_raw_parts(input_buffer, input_len_in) };

        Lexer {
            byte_offset: 0,
            token_offset: 0,
            byte_length: 1,
            token_length: 1,
            prev_byte_offset: 0,
            prev_token_offset: 0,
            peek_byte_offset: 0,
            peek_token_offset: 0,
            line: 0,
            token_type: 0,
            input: u8_slice,
        }
    }

    pub fn setToken(&mut self, type_in: i32, byte_length_in: u32, token_length_in: u32) -> i32 {
        self.token_type = type_in;
        self.byte_length = byte_length_in;
        self.token_length = token_length_in;
        return type_in;
    }

    pub fn get_byte_at(&self, index: usize) -> u8 {
        self.input[index]
    }

    pub fn codepoint(&mut self) -> u32 {
        let cp = get_utf8_code_point_at(self.byte_offset as usize, self.input);
        self.byte_length = get_utf8_byte_length_from_code_point(cp);
        self.token_length = get_token_length_from_code_point(cp);
        return cp;
    }

    pub fn class(&mut self) -> u32 {
        return getTypeAt(self.codepoint());
    }

    pub fn byte(&mut self) -> u32 {
        self.byte_length = 1;
        self.token_length = 1;
        return self.input[self.byte_offset as usize] as u32;
    }

    pub fn copy_in_place(&self) -> Lexer {
        let mut destination: Lexer = Lexer::new(self.input.as_ptr(), self.input.len());
        destination.peek_unroll_sync(self);
        return destination;
    }

    pub fn peek_unroll_sync(&mut self, source: &Lexer) {
        self.byte_offset = source.byte_offset;
        self.byte_length = source.byte_length;
        self.token_length = source.token_length;
        self.token_offset = source.token_offset;
        self.prev_byte_offset = source.prev_byte_offset;
        self.prev_token_offset = source.prev_token_offset;
        self.line = source.line;
        self.token_type = source.token_type;
    }

    pub fn scanner_sync(&mut self, source: &Lexer) {
        self.byte_offset = source.byte_offset;
        self.byte_length = source.byte_length;
        self.token_length = source.token_length;
        self.token_offset = source.token_offset;
        self.prev_byte_offset = self.byte_offset;
        self.prev_token_offset = self.byte_length;
        self.line = source.line;
        self.token_type = source.token_type;
    }

    pub fn sync_offsets(&mut self) {
        self.prev_byte_offset = self.byte_offset;
        self.prev_token_offset = self.token_offset;
    }

    pub fn set_token_span_to(&mut self, source: &Lexer) {
        if source.prev_byte_offset > self.byte_offset {
            self.byte_length = source.prev_byte_offset - self.byte_offset;
            self.token_length = source.prev_token_offset - self.token_offset;
        }
    }
    pub fn skip_delta(&mut self) -> u32 {
        return self.token_offset - self.prev_token_offset;
    }

    pub fn peek(&mut self) {
        self.byte_offset += self.byte_length as u32;
        self.token_offset += self.token_length as u32;
        if self.input.len() <= self.byte_offset as usize {
            self.token_type = 1;
            self.byte_length = 0;
            self.token_length = 0;
        } else {
            {
                if self.input[self.byte_offset as usize] == 10 {
                    self.line += 1
                };
                self.token_type = 0;
                self.byte_length = 1;
                self.token_length = 1;
            }
        };
    }

    pub fn next(&mut self) {
        self.peek();
        self.peek_byte_offset = self.byte_offset;
        self.peek_token_offset = self.token_offset;
    }

    pub fn consume(&mut self) {
        self.prev_byte_offset = self.byte_offset;
        self.prev_token_offset = self.token_offset;
    }

    pub fn reset(&mut self) {
        if self.byte_offset > self.peek_byte_offset {
            self.byte_offset = self.peek_byte_offset;
            self.token_offset = self.peek_token_offset;
            self.token_length = 1;
            self.byte_length = 1;
            self.token_type = 0;
        }
    }

    pub fn END(&mut self) -> bool {
        return self.byte_offset as usize >= self.input.len();
    }
}

fn getTypeAt(code_point: u32) -> u32 {
    (CHAR_LU_TABLE[code_point as usize] & 0x1F) as u32
}

File: packages/rt_go/recognize/lexer.go
package hc_kernel

// Kernel Lexer ----------------------------
type KernelLexer struct {
	input *[]uint8

	byte_offset       uint32
	token_offset      uint32
	token_length      uint32
	byte_length       uint32
	prev_byte_offset  uint32
	prev_token_offset uint32
	peek_byte_offset  uint32
	peek_token_offset uint32
	line              uint32
	token_type        uint32
	input_len         uint32
}

func newKernelLexer(input *[]uint8) *KernelLexer {

	p := KernelLexer{
		input:             input,
		byte_offset:       0,
		token_offset:      0,
		token_length:      1,
		byte_length:       1,
		prev_byte_offset:  0,
		prev_token_offset: 0,
		peek_byte_offset:  0,
		peek_token_offset: 0,
		line:              0,
		token_type:        0,
		input_len:         uint32(len(*input)),
	}

	return &p
}

func getTypeAt(code_point uint32) uint32 { return uint32(Jumptable[code_point] & 0x1F) }

func (s *KernelLexer) P_token_offset() uint32 {
	return s.token_offset
}

func (s *KernelLexer) P_token_length() uint32 {
	return s.token_length
}

func (s *KernelLexer) P_line() uint32 {
	return s.line
}

func (s *KernelLexer) codepoint() uint32 {
	cp := get_utf8_code_point_at(s.byte_offset, *s.input)
	s.byte_length = get_utf8_byte_length_from_code_point(cp)
	s.token_length = get_token_length_from_code_point(cp)
	return cp
}

func (s *KernelLexer) class() uint32 {
	return getTypeAt(s.codepoint())
}

func (s *KernelLexer) byte() uint32 {
	s.byte_length = 1
	s.token_length = 1
	return uint32((*s.input)[s.byte_offset])
}

func (s *KernelLexer) copy_in_place() *KernelLexer {
	destination := newKernelLexer(s.input)
	destination.peek_unroll_sync(s)
	return destination
}

func (s *KernelLexer) peek_unroll_sync(source *KernelLexer) {

	s.byte_offset = source.byte_offset
	s.byte_length = source.byte_length
	s.token_length = source.token_length
	s.token_offset = source.token_offset
	s.prev_byte_offset = source.prev_byte_offset
	s.prev_token_offset = source.prev_token_offset
	s.line = source.line
	s.token_type = source.token_type
}

func (s *KernelLexer) scanner_sync(source *KernelLexer) {
	s.byte_offset = source.byte_offset
	s.byte_length = source.byte_length
	s.token_length = source.token_length
	s.token_offset = source.token_offset
	s.prev_byte_offset = s.byte_offset
	s.prev_token_offset = s.byte_length
	s.line = source.line
	s.token_type = source.token_type
}

func (s *KernelLexer) peek() {
	s.byte_offset += s.byte_length
	s.token_offset += s.token_length

	if s.input_len <= s.byte_offset {
		s.token_type = 1
		s.byte_length = 0
		s.token_length = 0
	} else {

		if (*s.input)[s.byte_offset] == 10 {
			s.line += 1
		}

		s.token_type = 0
		s.byte_length = 1
		s.token_length = 1
	}
}

func (s *KernelLexer) reset() {
	if s.byte_offset > s.peek_byte_offset {
		s.byte_offset = s.peek_byte_offset
		s.token_offset = s.peek_token_offset
		s.token_length = 1
		s.byte_length = 1
		if s.token_type != 1 {
			s.token_type = 0
		}
	}
}

func (s *KernelLexer) skip_delta() uint32 {
	return s.token_offset - s.prev_token_offset
}

func (s *KernelLexer) next() {
	s.peek()
	s.peek_byte_offset = s.byte_offset
	s.peek_token_offset = s.token_offset
}

func (s *KernelLexer) consume() {
	s.prev_byte_offset = s.byte_offset
	s.prev_token_offset = s.token_offset
}

func (s *KernelLexer) END() bool {
	return s.byte_offset >= s.input_len
}
